src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,131 @@
+    private ListNode<T> head;
+    private ListNode<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        ListNode<T> node = tail;
+        ListNode<T> newNode = new ListNode<>(node, value, null);
+        tail = newNode;
+        if (head == null) {
+            head = newNode;
+        } else {
+            node.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        ListNode<T> node = getNode(index);
+        ListNode<T> newNode = new ListNode<>(node.prev, value, node);
+        if (node.prev != null) {
+            node.prev.next = newNode;
+        } else {
+            head = newNode;
+        }
+        node.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNode(index).data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        ListNode<T> node = getNode(index);
+        T oldData = node.data;
+        node.data = value;
+        return oldData;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        size--;
+        ListNode<T> node = getNode(index);
+        if (size == 0 && index == 0) {
+            head = null;
+            tail = null;
+        } else if (index == 0) {
+            head = head.next;
+            head.prev = null;
+        } else if (index == size) {
+            tail = tail.prev;
+            tail.next = null;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+        return node.data;
     }
 
     @Override
     public boolean remove(T object) {
+        ListNode<T> node = head;
+        for (int i = 0; i < size; i++) {
+            if (node.data == null && object == null
+                    || node.data != null && node.data.equals(object)) {
+                remove(i);
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Index " + index + " is out of bounds");
+        }
+    }
+
+    private ListNode<T> getNode(int index) {
+        ListNode<T> node = head;
+        for (int i = 0; i < index; i++) {
+            node = node.next;
+        }
+        return node;
```

Comment:
https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#dont-iterate-your-nodes-from-the-head-to-tail-or-vice-versa